examples/ode-integrator.dx

```diff
@@ -1,60 +1,179 @@
-'## Euler Integrator
-Using only generic vector operations.
-
-Time : Type = Float
-
--- State of Euler solver: (old state, old time, new state, new time)
-def Eulerstate (a:Type) : Type = (a & Time & a & Time)
-
-def eulerStep
-  (dict: VSpace a) ?=> (zt: Eulerstate a)
-  (dynamics: a -> Time -> a) (dt: Time)
-  : (Eulerstate a) =
-  (z_old, t_old, z, t) = zt
-  (z, t, z + dt .* dynamics z t, t + dt)
-
-def singleEuler (dict: VSpace a) ?=> (dynamics: a -> Time -> a)
-(zt:Eulerstate a) (t1: Time) (dt: Time) : (Eulerstate a) =
-  -- Take steps until we pass t1
-  snd $ withState zt \ztref.
-    while ( \().
-      (_, _, _, t) = get ztref
-      (t < t1)) \().  -- stopping condition
-        ztref := eulerStep (get ztref) dynamics dt
-
-def linear_interp
-  (dict: VSpace a) ?=> (z0: a) --o (z1: a) --o (t0: Time) (t1: Time) (t: Time) --o : a =
-  z0 + ((t - t0) / ((t1 - t0) + 0.000000001)) .* (z1 - z0)
-
-def odeint (dict: VSpace a) ?=> (dynamics: a -> Time -> a)
-  (z0: a) (t0: Time) (times: n=>Time) (dt: Time) : n=>a =
-  -- todo: assert / enforce that times is monotonically increasing
-  -- todo: never go past final time
-
-  init = (z0, t0, z0, t0)
-  
-  body = \i state.
-    t = times.i
-    (za, ta, zb, tb) = singleEuler dynamics state t dt
-    zt = linear_interp za zb ta tb t
-    ((za, ta, zb, tb), zt)
-  snd $ scan init body
+'Integrate systems of ordinary differential equations (ODEs) using the Dormand-Prince method for adaptive integration
+stepsize calculation.
+This version is a port of the [Jax implementation](https://github.com/google/jax/blob/4236eb2b5929b9643977553f7f988ca518b7df4e/jax/experimental/ode.py).
+A notable algorithmic difference is that it uses a lower-triangular
+matrix type for the Butcher tableua, and so avoid zero-padding everywhere.
+
+Time = Float
+
+-- Should these go in the prelude?
+def length (x: d=>Float) : Float = sqrt $ sum for i. sq x.i
+def (./) (x: d=>Float) (y: d=>Float) : d=>Float = for i. x.i / y.i
+def dot (_:VSpace v) ?=> (s:d=>Float) (vs:d=>v) : v = sum for j. s.j .* vs.j
```

Comment:
This one is already in the prelude as `vdot`. I think we could move `length` there, although maybe it would be clearer to call it `norm` or `norm2` just so that it's harder to confuse with `size`.

Reply:
The one in the prelude is
```
def vdot (x:n=>Float) (y:n=>Float) : Float = fsum \i. x.i * y.i
```
which I didn't want to convert to the generic vector version since then it couldn't use the faster ```fsum```.

Since ```fsum``` is slated for unification with ```sum```, how about I put my ```dot``` in the prelude, and we can delete ```vdot``` later when ```fsum``` is made obsolete?

Reply:
Oh, we should remove `fsum`. I've already implemented the inlining optimization necessary to make `sum` just as fast!